Code/Framework/AzToolsFramework/AzToolsFramework/ComponentMode/ComponentModeSwitcher.cpp

```diff
@@ -0,0 +1,381 @@
+                });
+            componentDataIt == m_addedComponents.end())
+        {
+            AddSwitcherButton(newComponentData, newComponentData.m_componentName.c_str(), newComponentData.m_iconPath.c_str());
+            m_addedComponents.push_back(newComponentData);
+        }
+    }
+
+    void ComponentModeSwitcher::AddSwitcherButton(ComponentData& componentData, const char* componentName, const char* iconPath)
+    {
+        ViewportUi::ButtonId buttonId;
+        ViewportUi::ViewportUiRequestBus::EventResult(
+            buttonId,
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::CreateSwitcherButton,
+            m_switcherId,
+            iconPath,
+            componentName);
+
+        componentData.m_buttonId = buttonId;
+
+        AZStd::string buttonTooltip = AZStd::string::format(" Enter %s component mode", componentData.m_componentName.c_str());
+
+        ViewportUi::ViewportUiRequestBus::Event(
+            ViewportUi::DefaultViewportId,
+            &ViewportUi::ViewportUiRequestBus::Events::SetSwitcherButtonTooltip,
+            m_switcherId,
+            buttonId,
+            buttonTooltip);
+    }
+
+    void ComponentModeSwitcher::RemoveComponentButton(const AZ::EntityComponentIdPair pairId)
+    {
+        // find pairId in m_whatever, call bus to remove
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [pairId](const ComponentData& componentInfo)
+                {
+                    return componentInfo.m_pairId == pairId;
+                });
+            componentDataIt != m_addedComponents.end())
+        {
+            ViewportUi::ViewportUiRequestBus::Event(
+                ViewportUi::DefaultViewportId,
+                &ViewportUi::ViewportUiRequestBus::Events::RemoveSwitcherButton,
+                m_switcherId,
+                componentDataIt->m_buttonId);
+
+            m_addedComponents.erase(componentDataIt);
+        }
+    }
+
+    void ComponentModeSwitcher::ActivateComponentMode(const ViewportUi::ButtonId buttonId)
+    {
+        bool inComponentMode;
+        AzToolsFramework::ComponentModeFramework::ComponentModeSystemRequestBus::BroadcastResult(
+            inComponentMode, &ComponentModeSystemRequests::InComponentMode);
+        ComponentData* component = nullptr;
+
+        // find component associated with button Id
+        if (auto componentDataIt = AZStd::find_if(
+                m_addedComponents.begin(),
+                m_addedComponents.end(),
+                [buttonId](const ComponentData& componentData)
+                {
+                    return componentData.m_buttonId == buttonId;
+                });
+            componentDataIt != m_addedComponents.end())
+        {
+            component = componentDataIt;
+        }
```

Comment:
update: We only actually use `component` inside the `buttonId != m_transformButtonId` block, so we could move this code there (again closer to first use)

Reply:
I put the whole thing in the if block because there were things that needed to be done inbetween the component being set and the `buttonId != m_transformButtonId` block

Reply:
Ah okay I see, is that still the case now?